src/logql-query.ts

```diff
@@ -1,48 +1,250 @@
+        }
+
+        return false;
+      }
+    },
+    leave(selectorsNode) {
+      if (selectorsNode.name === 'PipelineStage') {
+        pipelineStages.push({ operator, value, labelsInFilter });
+        labelsInFilter = undefined;
+      }
+    },
+  });
+
+  return pipelineStages;
+};
 
 export class LogQLQuery {
-  selectors: Array<string> = [];
-  pipeline: Array<string> = [];
+  streamSelector: Array<LabelMatcher> = [];
+  pipeline: Array<PipelineStage> = [];
 
   constructor(query: string) {
-    const parts = query.match(/.*{([^}]*)}(.*)/);
-    this.selectors =
-      parts && parts[1].length > 0
-        ? parts[1].split(',').filter(notEmptyString).map(trim)
-        : [];
-    this.pipeline =
-      parts && parts[2].length > 0
-        ? parts[2].split('|').filter(notEmptyString).map(trim)
-        : [];
+    const syntaxTree = parser.parse(query);
+    let parsedMatchers: Array<LabelMatcher> = [];
+    let parsedPipeline: Array<PipelineStage> = [];
+
+    syntaxTree.iterate({
+      enter(node) {
+        if (node.name) {
+          if (node.name === 'Selector' && parsedMatchers.length === 0) {
+            parsedMatchers = parseMatchers(syntaxTree, node, query);
+
+            return false;
+          } else if (
+            node.name === 'PipelineExpr' &&
+            parsedPipeline.length === 0
+          ) {
+            parsedPipeline = parsePipelineStages(syntaxTree, node, query);
+
+            return false;
+          }
+        }
+      },
+    });
+
+    this.streamSelector = parsedMatchers;
+    this.pipeline = parsedPipeline;
   }
 
-  appendSelector = (newSelectors: string | Array<string>) => {
-    if (Array.isArray(newSelectors)) {
-      this.selectors.push(...newSelectors);
-    } else {
-      this.selectors.push(newSelectors);
+  removeSelectorMatcher = (matcherToRemove: Partial<LabelMatcher>) => {
+    this.streamSelector = this.streamSelector.filter((matcher) => {
+      let include = true;
+
+      if (matcherToRemove.label) {
+        include = include && matcher.label !== matcherToRemove.label;
+      }
+
+      return include;
+    });
+
+    return this;
+  };
+
+  addSelectorMatcher = (matchers?: LabelMatcher | Array<LabelMatcher>) => {
+    if (!matchers) {
+      return this;
     }
 
-    this.selectors = this.selectors.filter(notEmptyString).map(trim);
+    const matchersArray = Array.isArray(matchers) ? matchers : [matchers];
+
+    matchersArray.forEach((matcher) => {
+      const existingSelectorIndex = this.streamSelector.findIndex(
+        ({ label }) => matcher.label === label,
+      );
+
+      if (existingSelectorIndex !== -1) {
+        this.streamSelector[existingSelectorIndex] = matcher;
+      } else {
+        this.streamSelector.push(matcher);
+      }
+    });
+
+    this.streamSelector = this.streamSelector.filter(
+      (selector, index) =>
+        this.streamSelector.findIndex(
+          ({ label }) => selector.label === label,
+        ) === index,
+    );
```

Comment:
I looks to me like this filtering is redundant because we check for a matching label before adding the matcher to `this.streamSelector`.